cli/pcluster/update/update_command.py

```diff
@@ -0,0 +1,151 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+# with the License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import print_function
+
+import logging
+import sys
+import time
+
+import boto3
+from botocore.exceptions import ClientError
+from tabulate import tabulate
+
+import pcluster.utils as utils
+from pcluster.config.config_patch import ConfigPatch
+from pcluster.config.pcluster_config import PclusterConfig
+
+LOGGER = logging.getLogger(__name__)
+
+
+def execute(args):
+    LOGGER.info("Validating configuration file {0}...".format(args.config_file))
+    stack_name = utils.get_stack_name(args.cluster_name)
+    target_config = PclusterConfig(
+        config_file=args.config_file, cluster_label=args.cluster_template, fail_on_file_absence=True
+    )
+    target_config.validate()
+    cfn_params = target_config.to_cfn()
+    cfn = boto3.client("cloudformation")
+    _autofill_cfn_params(cfn, args, cfn_params, stack_name, target_config)
+
+    if _check_changes(args, target_config):
+        _update_cluster(args, cfn, cfn_params, stack_name)
+    else:
+        LOGGER.info("Update aborted.")
+        sys.exit(1)
+
+
+def _update_cluster(args, cfn, cfn_params, stack_name):
+    LOGGER.info("Updating: %s", args.cluster_name)
+    LOGGER.debug("Updating based on args %s", str(args))
+    try:
+        LOGGER.debug(cfn_params)
+        if args.extra_parameters:
+            LOGGER.debug("Adding extra parameters to the CFN parameters")
+            cfn_params.update(dict(args.extra_parameters))
+
+        cfn_params = [{"ParameterKey": key, "ParameterValue": value} for key, value in cfn_params.items()]
+        LOGGER.info("Calling update_stack")
+        cfn.update_stack(
+            StackName=stack_name, UsePreviousTemplate=True, Parameters=cfn_params, Capabilities=["CAPABILITY_IAM"]
```

Comment:
Would it be possible to use
```python
      template_url = _evaluate_pcluster_template_url(pcluster_config)
        cfn.update_stack(
            StackName=stack_name, TemplateURL=template_url, Parameters=cfn_params, Capabilities=["CAPABILITY_IAM"]
        )
```
safely here? 

My motivation is for this si that I've set up cfn-hup on my cluster, but rely on incrementing an environment variable in the MasterServer's UserData to trigger the update. That means I can't use pcluster update, as it re-uses the same stack template.

Reply:
Hi @joeydumont, 

We disabled this possibility to avoid to pass wrong cloudformation templates (or template associated to a different CLI version) when updating a cluster and be sure to preserve the cluster functionality.

A workaround for your case could be to create a custom template, on which the extra_json parameter will be used/saved somewhere in the user-data of the master instance.

Once you create a cluster with this custom template you'll be able to trigger the cfn-hup by updating the cluster after a change in the extra_json configuration parameter.

BTW I'd suggest to create a new github issue for this interesting question instead of commenting in the PR itself. It could be useful to start a discussion with other users and we'll be sure to not lose it.

Let us know if the workaround works for you.

Reply:
Thanks @enrico-usai, I've opened #1639. It's about more than use `UsePreviousTemplate` though. 

I've used a similar workaround. Because of `cfn-hup`, I ended up doing the same thing that you did in #1564 in my custom template. Using `extra_json` works for now, but I would like to be able to choose to use `TemplateURL` in `pcluster update` nonetheless.

Thanks for the reply!